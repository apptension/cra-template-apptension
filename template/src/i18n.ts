import enTranslationMessages from './translations/en.json';
import plTranslationMessages from './translations/pl.json';

export const LOCALES = {
  ENGLISH: 'en',
  POLISH: 'pl',
};

type MessagesObject = { [key: string]: string };

export const DEFAULT_LOCALE = LOCALES.ENGLISH;

export const appLocales = [LOCALES.ENGLISH, LOCALES.POLISH];

export const formatTranslationMessages = (locale: string, messages: MessagesObject): MessagesObject => {
  const defaultFormattedMessages: MessagesObject =
    locale !== DEFAULT_LOCALE ? formatTranslationMessages(DEFAULT_LOCALE, enTranslationMessages) : {};
  return Object.keys(messages).reduce((formattedMessages, key) => {
    const formattedMessage =
      !messages[key] && locale !== DEFAULT_LOCALE ? defaultFormattedMessages[key] : messages[key];
    return Object.assign(formattedMessages, { [key]: formattedMessage });
  }, {});
};

export const translationMessages = {
  [LOCALES.ENGLISH]: formatTranslationMessages(LOCALES.ENGLISH, enTranslationMessages),
  [LOCALES.POLISH]: formatTranslationMessages(LOCALES.POLISH, plTranslationMessages),
};
