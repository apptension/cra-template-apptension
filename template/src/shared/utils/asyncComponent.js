import loadable from '@loadable/component';

export const asyncComponent = (asyncLoader, exportedAs) => {
  return loadable(async () => {
    const module = await asyncLoader();
    return exportedAs ? module[exportedAs] : module;
  });
};
