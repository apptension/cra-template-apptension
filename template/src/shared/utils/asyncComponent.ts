import loadable from '@loadable/component';

export const asyncComponent = (asyncLoader: Function, exportedAs?: string) => {
  return loadable(async () => {
    const module = await asyncLoader();
    return exportedAs ? module[exportedAs] : module;
  });
};
