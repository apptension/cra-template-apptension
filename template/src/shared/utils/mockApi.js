import nock from 'nock';

export default nock('http://localhost');
