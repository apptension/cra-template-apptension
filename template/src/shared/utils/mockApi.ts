import nock from 'nock';

const baseUrl = process.env.REACT_APP_BASE_API_URL;

if (!baseUrl) {
  throw new Error('REACT_APP_BASE_API_URL env is missing');
}

export const mockApi = nock(baseUrl);
