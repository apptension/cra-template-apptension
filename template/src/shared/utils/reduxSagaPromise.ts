import { useDispatch } from 'react-redux';
import {
  ActionCreatorWithPayload,
  ActionCreatorWithPreparedPayload,
  createAction,
  Middleware,
  PayloadAction,
  PrepareAction,
} from '@reduxjs/toolkit';

import { put } from 'redux-saga/effects';
import { GlobalState } from '../../config/reducers';

export interface PromiseActionMeta<A, B> {
  promise: {
    resolveAction: ActionCreatorWithPayload<A>;
    rejectAction: ActionCreatorWithPayload<B>;
    resolve?(value: A): void;
    reject?(value: B): void;
  };
}

export interface PromiseActionCreatorWithPayload<P, A, B = void> {
  (payload: P): PromiseAction<P, A, B>;
  type: string;
  trigger: ActionCreatorWithPreparedPayload<[P], PreparePromiseAction<P, A, B>>;
  resolved: ActionCreatorWithPayload<A>;
  rejected: ActionCreatorWithPayload<B>;
}

export type PreparePromiseAction<P, A, B> = (
  ...args: [P]
) => {
  payload: P;
  meta: PromiseActionMeta<A, B>;
};

export type PromiseAction<P = void, A = void, B = void> = PayloadAction<P, string, PromiseActionMeta<A, B>>;

export function createActionRoutine<P = void, A = void, B = void>(
  prefix: string,
  meta?: any
): PromiseActionCreatorWithPayload<P, A, B> {
  const resolveAction = createAction<PrepareAction<A>, string>(`${prefix}.RESOLVED`, (payload) => ({ payload }));
  const rejectAction = createAction<PrepareAction<B>>(`${prefix}.REJECTED`, (payload) => ({ payload }));

  const triggerAction = createAction<PreparePromiseAction<P, A, B>>(`${prefix}.TRIGGER`, (payload: P) => ({
    payload,
    meta: {
      ...(meta || {}),
      promise: {
        resolveAction,
        rejectAction,
      },
    },
  }));

  return ((): PromiseActionCreatorWithPayload<P, A, B> => {
    const f: any = triggerAction;
    f.trigger = triggerAction;
    f.resolved = resolveAction;
    f.rejected = rejectAction;
    return f;
  })();
}

export function createPromiseAction<P = void, A = void, B = void>(prefix: string) {
  return createActionRoutine<P, A, B>(prefix, { promisified: true });
}

export function* resolvePromiseAction<P = void, A = void, B = void>(action: PromiseAction<P, A, B>, value: A) {
  yield put(action.meta.promise.resolveAction(value));
  action.meta?.promise?.resolve?.(value);
}

export function* rejectPromiseAction<P = void, A = void, B = void>(action: PromiseAction<P, A, B>, value: B) {
  yield put(action.meta.promise.rejectAction(value));
  action.meta?.promise?.reject?.(value);
}

const isPromiseTriggerAction = (action: PayloadAction<any, string, any>) =>
  !!action.meta?.promise?.resolveAction && !!action.meta?.promisified;

export const promiseMiddleware: Middleware<unknown, GlobalState> = (store) => (next) => (
  action: PayloadAction<any, string, any>
) => {
  if (isPromiseTriggerAction(action)) {
    return new Promise((resolve, reject) =>
      next({
        ...action,
        meta: {
          promise: {
            ...action.meta.promise,
            resolve,
            reject,
          },
        },
      })
    );
  }

  return next(action);
};

type PromiseDispatch = <P, A, B>(action: PromiseAction<P, A, B>) => Promise<A>;
export const useAsyncDispatch = () => useDispatch<PromiseDispatch>();
