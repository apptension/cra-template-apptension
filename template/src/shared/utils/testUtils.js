import { DEFAULT_LOCALE } from '../../i18n';

export const intlMock = () => ({
  formatMessage: ({ id, defaultMessage, values = {} }) => `${id} / ${defaultMessage} / ${JSON.stringify(values)}`,
  locale: DEFAULT_LOCALE,
  formatNumber: (value, options) => `${value} / ${JSON.stringify(options)}`,
});
