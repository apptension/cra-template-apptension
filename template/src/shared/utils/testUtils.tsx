import React, { ReactNode, ReactElement, FC } from 'react';
import { HelmetProvider } from 'react-helmet-async';
import { render } from '@testing-library/react';
import { createStore } from 'redux';
import { identity } from 'ramda';
import { Provider } from 'react-redux';
import { createMemoryHistory, MemoryHistory } from 'history';
import { Route, Router } from 'react-router';
import { IntlProvider } from 'react-intl';
import produce from 'immer';

import { DEFAULT_LOCALE, translationMessages, MessagesObject } from '../../i18n';
import { store as fixturesStore } from '../../../fixtures/store';
import { ResponsiveThemeProvider } from '../components/responsiveThemeProvider';
import { theme as defaultTheme } from '../../theme/theme';
import { GlobalState } from '../../modules/reducers';
import { LOCALES_INITIAL_STATE } from '../../modules/locales';
import { STARTUP_INITIAL_STATE } from '../../modules/startup';
import { USERS_INITIAL_STATE } from '../../modules/users';
//<-- IMPORT MODULE STATE -->

const defaultGlobalState: GlobalState = {
  locales: LOCALES_INITIAL_STATE,
  startup: STARTUP_INITIAL_STATE,
  users: USERS_INITIAL_STATE,
  //<-- INJECT MODULE STATE -->
};

export const PLACEHOLDER_TEST_ID = 'content';
export const PLACEHOLDER_CONTENT = <span data-testid="content">content</span>;

export const spiedHistory = (route = '/') => {
  const history = createMemoryHistory({ initialEntries: [route] });
  const pushSpy = jest.spyOn(history, 'push');
  return {
    history,
    pushSpy,
  };
};

interface ContextData {
  router?: {
    url?: string;
    routePath?: string;
    history?: MemoryHistory;
  };
  store?: GlobalState;
  messages?: MessagesObject;
  theme?: object;
}

interface ProvidersWrapperProps {
  children: ReactNode;
  context: ContextData;
}

export const ProvidersWrapper: FC<ProvidersWrapperProps> = ({ children, context = {} }) => {
  const { router = {}, store = fixturesStore, messages, theme = defaultTheme } = context;
  const { url = `/${DEFAULT_LOCALE}`, routePath = '/:lang/', history } = router;

  const routerHistory: MemoryHistory = history ?? createMemoryHistory({ initialEntries: [url] });

  const intlProviderMockProps = {
    locale: DEFAULT_LOCALE,
    messages: messages ?? translationMessages[DEFAULT_LOCALE],
  };

  return (
    <Router history={routerHistory}>
      <ResponsiveThemeProvider theme={theme}>
        <HelmetProvider>
          <IntlProvider {...intlProviderMockProps}>
            <Provider store={createStore(identity, store)}>
              <Route path={routePath}>{children}</Route>
            </Provider>
          </IntlProvider>
        </HelmetProvider>
      </ResponsiveThemeProvider>
    </Router>
  );
};

export const makeContextRenderer = <T, _>(component: (props: T | {}) => ReactElement) => (
  props?: T,
  context?: ContextData
) =>
  render(component(props ?? {}), {
    wrapper: ({ children }) => <ProvidersWrapper context={context}>{children}</ProvidersWrapper>,
  });

export const makePropsRenderer = <T, _>(component: (props: T | {}) => ReactElement) => (props?: T) =>
  render(component(props ?? {}));

export const prepareState = (stateSetter: (draftState: GlobalState) => void) =>
  produce(defaultGlobalState, stateSetter);
