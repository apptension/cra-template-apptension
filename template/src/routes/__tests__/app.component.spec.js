import React from 'react';
import { makePropsRenderer } from '../../shared/utils/testUtils';
import { AppComponent } from '../app.component';
import { DEFAULT_LOCALE } from '../../i18n';

const defaultProps = {
  children: <div className="app__children">Children</div>,
  language: DEFAULT_LOCALE,
};

describe('App: Component', () => {
  const component = props => <AppComponent {...defaultProps} {...props} />;
  const render = makePropsRenderer(component);

  it('should render App when language is set', () => {
    const { container } = render();
    expect(container).toMatchSnapshot();
  });
});
