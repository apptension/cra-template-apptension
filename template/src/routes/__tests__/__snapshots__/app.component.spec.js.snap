// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App: Component should not render App when language is not set 1`] = `""`;

exports[`App: Component should render App when language is set 1`] = `
<IntlProvider
  key="en"
  locale="en"
  messages={
    Object {
      "app.pageTitle": "Apptension Boilerplate",
      "home.pageTitle": "Homepage",
      "home.welcome": "Hello World!",
      "notFound.pageTitle": "Not found",
      "notFound.title": "Error: 404",
      "unsupported.pageTitle": "Unsupported Browser",
      "unsupported.title": "Unsupported Browser",
    }
  }
>
  <ResponsiveThemeProvider
    theme={
      Object {
        "color": Object {
          "white": "#fffffff",
        },
        "font": Object {
          "primary": "OpenSans",
        },
      }
    }
  >
    <FormattedMessage
      defaultMessage="Apptension Boilerplate"
      id="app.pageTitle"
    >
      <Component />
    </FormattedMessage>
    <GlobalStyleComponent />
    <div
      className="app__children"
    >
      Children
    </div>
  </ResponsiveThemeProvider>
</IntlProvider>
`;
